Change information stored in ` _topic_node_index` to avoid oversized alloc

ballard26 opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
 ` _topic_node_index` originally stored a bunch of metadata about every group a node led. However, this information was only ever used in `even_topic_distributon_constraint::recommended_reassignment`. A method which isn't used anywhere outside of unit tests.

This PR changes the information stored in `_topic_node_index` to be just the count of leaders on a given node. And removes the implementation of `even_topic_distributon_constraint::recommended_reassignment`.

Fixes #17349 
## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v23.3.x
- [ ] v23.2.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/46641#018e685e-435f-491b-bfd0-b1d51343add3
<!--02463bb3-ff16-489d-94a7-ff3b815d85a6-->

ballard26 commented:
> I like all the code removal, but can you elaborate on why you decided to take this route over just changing the `std::vector` to `chunked_vector`?

Sure, initially I did just blindly switch it over. Then I realized `chunked_vector` doesn't have a `delete(iterator ..)` implementation. So the option was to write an implementation of that or do something else. And even if I had switched over the balancer would still be wasting a lot of time and memory book keeping information it never uses.

I implemented `recommend_reassignment` for the topic aware constraint with the idea that it could be used as a speed improvement over just randomly selecting reassignments and seeing if they decrease error. In testing it did prove to more quickly decrease error. However, I wasn't able to come up with a good way to combine the `recommend_reassignment` methods from all constraints in a way that reduced error as far as just randomly selecting and checking reassignments. So it never actually was put into use. I still have the data and graphs I generated during the investigation back then if you're interested.

So with that in mind `recommend_reassignment` is basically dead code as is. And the only thing its doing is creating more book keeping work for the balancer.

ballard26 commented on src/v/cluster/scheduling/leader_balancer_constraints.cc:
```diff
@@ -64,85 +52,7 @@ void even_topic_distributon_constraint::update_index(const reassignment& r) {
-    }
-
-    return std::nullopt;
+    vassert(false, "not implemented");
```
Will do

ballard26 replied:
Done

BenPope replied:
@ballard26 this seems to have introduced a CI failure: https://buildkite.com/redpanda/redpanda/builds/51764#0190c949-1a0f-42ba-a040-4e33bc85fc38

dotnwat replied:
> ERROR 2024-07-19 07:45:03,854 [shard 0:main] assert - Assert failure: (/var/lib/buildkite-agent/builds/buildkite-amd64-builders-i-0269b6a11fdecc85f-1/redpanda/redpanda/src/v/cluster/scheduling/leader_balancer_constraints.cc:59) 'false' not implemented
 
thanks @BenPope looks like something changed, or the assumptions weren't correct initially.

dotnwat replied:
I searched a little but didn't see a ticket for this. is there one @BenPope?

ballard26 commented on src/v/cluster/scheduling/leader_balancer_constraints.cc:
```diff
@@ -42,20 +42,8 @@ void even_topic_distributon_constraint::update_index(const reassignment& r) {
-    _topic_node_index[topic_id][r.to.node_id].emplace_back(
-      moved_group_info.group_id, r.to, std::move(moved_group_info.replicas));
+    _topic_node_index.at(topic_id).at(r.from.node_id) -= 1;
+    _topic_node_index.at(topic_id).at(r.to.node_id) += 1;
```
It should always be. `even_topic_distributon_constraint::update_index` is only used internally and the reassignment is validated by the callers before its called.

ztlpn replied:
Got it, we initialize it on (new) line 79. But maybe it's easier for the reader to just use `operator[]`

ballard26 commented on src/v/cluster/scheduling/leader_balancer_constraints.h:
```diff
@@ -171,8 +157,7 @@ class even_topic_distributon_constraint final
 
-    topic_map<absl::flat_hash_map<model::node_id, std::vector<group_info>>>
-      _topic_node_index;
+    topic_map<absl::flat_hash_map<model::node_id, size_t>> _topic_node_index;
```
Good point, adding a comment.

piyushredpanda commented:
/ci-repeat

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/47424#018eac40-c414-4630-a9ad-be99d0604a0e:
```
"rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes"
```
<!--596583d0-14f5-4c75-84ef-18d558b7c907-->

ballard26 commented:
/ci-repeat

piyushredpanda commented:
Failures should be fixed by https://github.com/redpanda-data/redpanda/pull/18302

vbotbuildovich commented:
/backport v23.3.x

piyushredpanda commented:
/backport v24.1.x
